natpmp/client.go

```diff
@@ -0,0 +1,204 @@
+package natpmp
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+// Version is the expected protocol version for NAT-PMP.
+const Version = 0
+
+// ErrProtocol indicates that a NAT gateway returned a response that violates
+// the NAT-PMP protocol.
+var ErrProtocol = errors.New("natpmp: protocol error")
+
+//go:generate stringer -type=Error -output=strings.go
+
+// An Error is a NAT-PMP protocol result code which indicates that an operation
+// has failed.
+type Error int
+
+// Possible Errors as defined in RFC 6886, section 3.5.
+const (
+	UnsupportedVersion Error = 1
+	NotAuthorized      Error = 2
+	NetworkFailure     Error = 3
+	OutOfResources     Error = 4
+	UnsupportedOpcode  Error = 5
+
+	// Although success is technically a result code, we don't expose it
+	// directly to the user because a successful operation returns nil error.
+	success Error = 0
+)
+
+// Error implements error.
+func (e Error) Error() string {
+	return fmt.Sprintf("natpmp: result %d: %s", e, e.String())
+}
+
+// A Client is a NAT-PMP client which can communicate with a NAT gateway.
+type Client struct {
+	// The UDP socket and address used to communicate with a NAT gateway.
+	mu      sync.Mutex
+	pc      net.PacketConn
+	gateway net.Addr
+}
+
+// Dial creates a Client which communicates with the NAT gateway specified by
+// addr.
+func Dial(addr string) (*Client, error) {
```

Comment:
There really needs to be an API shape here that lets callers provide their own PacketConn for sending, and provide their own raw packets to receive. For NAT-PMP and PCP to work, you have to use the same socket you'd use for your "main" protocol, so in practice the caller of this library is going to have a receive demultiplexer that identifies NAT-PMP packets and passes them back to this library - rather than this library reading from the PacketConn itself.

Reply:
For some inspiration on this, you can take a look at the netcheck package in tailscale/tailscale. It does STUN and some UPnP/NAT-PMP/PCP discovery, and uses this pattern of "here's a PacketConn for sending, and here's a function call for you to give me packets that look like they're for me".

Reply:
Will do.